kasa/smartstrip.py

```diff
@@ -227,9 +238,21 @@ async def _query_helper(
         self, target: str, cmd: str, arg: Optional[Dict] = None, child_ids=None
     ) -> Any:
         """Override query helper to include the child_ids."""
-        return await self.parent._query_helper(
-            target, cmd, arg, child_ids=[self.child_id]
-        )
+        try:
+            return await self.parent._query_helper(
+                target, cmd, arg, child_ids=[self.child_id]
+            )
+        except Exception as ex:
+            if len(self.child_id) < 3:
+                # the discovery protocol only returns the index for the child ID
+                #  but the 2021 KP400 requires the full ID including the parent ID
+                #  so we swallow this exception, update the child_id and try again
+                self.child_id = self.parent.sys_info["deviceId"] + self.child_id
+                return await self.parent._query_helper(
+                    target, cmd, arg, child_ids=[self.child_id]
+                )
+            else:
+                raise ex
```

Comment:
Could you please add a test for this behavior, just so that it won't get broken at some point if this gets re factored/changed?

Reply:
working on it. This is requiring some deep dives into the testing code since the responsibility in FakeTransportProtocol leverages a bit of the underlying device so I need to somehow pull out the discovery data without impacting how the rest of the FakeTransportProtocol works and avoiding any major refactor of the testing code.

Reply:
As a warning, the code on those parts is quite old and likely fragile, so it might be simpler just to refactor (or rewrite, or simply mock) the necessary parts inside `test_smartstrip.py` :-)

Reply:
So, we still want to have some sort of testing for this code path. My proposal would be the following:
1. Add a separate test that mocks `self.parent._query_helper` to raise an exception
2. Verify that the mock gets called once for a sysinfo (which can be injected directly in the test, no need to touch the fixture files)
3. Verify that the mock gets called twice if the sysinfo has a short-form id.

Reply:
added a test in test_strip. 
(1) modified the fake transport to throw an exception if there are child_id passed that it did not use. 
(2) changed the child_id for the plug to simulate the discovery filling in only the last two characters of the index
(3) then checked to make sure that turn on and turn off still worked and that the code restored the full child_id including the expected device_id prefix

Reply:
@rytilahti  - let me know what you think